Add information about Secure Storage in Studio 7.11 (#267)
mulesoft/docs-studio: modules/ROOT/pages/secure-storage.adoc

> = Secure Storage Overview

Reviewer: BTW, we should probably start calling these "pages" rather than "articles" (or "nodes" in the left nav.), per the nomenclature proposed in the templates initiative.

Reviewer: Also, this is not really an overview, because it has several very complete subtopics, including two tasks.
Did you intend for the tasks to have H1 titles?
Otherwise, this page should be retitled to imply audience participation, not just brief overview points.

> Anypoint Studio 7.10 and earlier versions stored the information of authenticated users in the workspace, meaning that you could have different authenticated users per workspace.

Reviewer: ```suggestion
Anypoint Studio 7.10 and earlier store user authentication in the workspace, meaning that each workspace has its own unique set of authorized users.
```
Are we also implying that any one user might have different personal credentials based on workspace rather than on the user themself?

Author: @kbostrom 

> (...) each workspace has its own unique set of authorized users.

I think this changes the meaning of the original sentence.

What do you think about?
> Anypoint Studio 7.10 and earlier store user authentication in the workspace, meaning that each workspace can contain different user authentications.

> Studio 7.11 introduces a Secure Storage facility to encrypt and save your credentials.

Reviewer 2: I would make it clear that the Secure Storage it isn't developed by us. Maybe something like:

- Studio 7.11 now uses the Secure Storage facility provided by Eclipse to encrypt and save your credentials.
- Studio 7.11 introduces the usage of the Secure Storage facility provided by Eclipse to encrypt and save your credentials.

> Studio 7.11 introduces a Secure Storage facility provided by Eclipse to encrypt and save your credentials.

Reviewer: ```suggestion
Studio 7.11 introduces Secure Storage, provided by Eclipse to encrypt and save your credentials.
```
"...credentials across all workspaces."?

Author: I like this. I was hesitant about giving it its own feature name (name in capital letters), but I see other Eclipse-based applications doing something similar.

> Studio provides a master password provider that automatically generates a master password for all your user credentials. When adding a new credential, Studio uses your master password to encrypt and save that credential into a secure storage file in your file system.

Reviewer: ```suggestion
Studio now includes a master password provider that automatically generates one master password per user, used to authenticate all of that user's credentials. And when a user adds a new credential, Studio uses their master password to encrypt and save it into a secure storage file in their file system.
```
I would have preferred to use "you" rather than "user," but I could not figure out how; maybe you can?

Author: > one master password per user, used to authenticate all of that user's credentials

Not "authenticate", but "encrypt". I'll copy/paste your suggestion and change that word.

Reviewer: Or...have I misunderstood: is STUDIO adding the new credential?
```suggestion
Studio now includes a master password provider that automatically generates one master password per user, used to authenticate all of that user's credentials. When Studio subsequently adds a new credential, it uses this master password to encrypt and save that new credential into a secure storage file in the user's file system.
```

> Because of the architecture of the secure storage, the authenticated users are now shared across all Eclipse-based workspaces and Anypoint Studio instances.

Reviewer: ```suggestion
The architecture of Secure Storage enables user authentication to be shared across all Eclipse-based workspaces and Anypoint Studio instances.
```

Reviewer: The above suggestions are just that: suggestions. You can likely develop your own adjusted wording. 
What I would like to add here is that I think this text might be condensed further and made even simpler by using an intro something like...:

Secure Storage in Studio 7.11

Anypoint Studio 7.11 introduces Secure Storage, a new architecture in which all of your credentials across Eclipse-based workspaces and Studio instances use a single master password. With Secure Storage, your credentials are encrypted and saved to one secure storage file in your own filesystem rather than scattered across working environments.

Author: We don't want to make this a "Studio" feature, because we  don't have control over it. We inherit this from Eclipse, and the user has no choice in whether using this or not.  

At some point I wanted to start the page with something like this:

> Because Anypoint Studio is an Eclipse-based application, it relies on the underlying existing architecture of the Eclipse platform. Therefore, Studio re-uses the Eclipse secure storage provided mechanism.

It might be necessary to convey the idea that this is how Studio works now and it should not bother the user that much.  Still, because it's a behavior change, it has to be documented. What do you think @kbostrom ?

== Master Password Providers

> The available master password providers depend on your operative system. Studio assigns a priority to the available master password providers and enables you to choose which one to use. By default, Studio assigns a higher level of priority to the password provider that uses your OS's underlying capabilities to generate and encrypt your master password.

Reviewer 2: Maybe I would be less assertive that it is Studio who do things. It is really Eclipse, we just use the default providers and behavior defined by Eclipse.

> The available master password providers depend on your operative system. Studio assigns a priority to the available master password providers and enables you to choose which one to use. By default, Studio assigns a higher level of priority to the password provider that uses your OS's underlying capabilities to generate and encrypt your master password.

Reviewer: I'm not sure that we need this paragraph...or at least that we need the priority-setting text right up front.
I suggest this organization:

1. The master password provider available to you depends on your operating system:
2. list of systems: Windows, mac, Linux, and UI
3. When (you or Studio?) first create or add to your authorization credentials, Studio notifies you that your provider has automatically supplied a master password (and tells you what it is?)
4. Studio prompts you to create two hints to enable master password recovery
5. Studio caches your master password until (you manually clear it from cache or...close Studio? does this mean, then, that the master password changes every time you open Studio?) 

(I don't think we need the text about Studio prioritizing providers; it seems to simply choose the one that applies to your environment...or you automatically choose the one that best suits your environment.)

Reviewer: Be sure to change "operative" to "operating."

Author: Studio inherits this from Eclipse. The password does not change until the user chooses to re-generate it, and the password is never shown to the user. 
This is a behavior inherited from Eclipse and the user has no control over it, except when choosing to ignore the prioritization that "Studio" (Eclipse actually) applies to each password provider. -Which in my experience is changing from the inherited OS dependant provider to the UI where the user can define a password of his own.-

The term "password hint" seems broken as well. Those are actually security questions in case the master password is changed by another instance while the user still has the current password cached. 

@mdaloia Am I misunderstanding something about this behavior?

Reviewer 2: Hi, the behavior and config options as Fer described are correct.

> When adding a new credential to Studio, it automatically uses the master password generated by the password provider with the highest priority to encrypt your credential, and stores it in a secure storage file located in your filesystem.

Reviewer: Who or what is "it" that automatically uses...?
Also, clarify the "it" that is being stored.

Windows Integration (64 bit)::

> Available on Windows. This provider uses a random value that is encrypted based on your Windows login information and stored in secure storage. This provider uses the master password automatically to decrypt your credentials.

Reviewer: If the value is random, how can it be "based" on your login information?
Do we mean simply "a value based on your login information"?
```suggestion
Available on Windows. This provider creates and encrypts a master password based on your Windows login information and stores it in secure storage. Windows Integration then uses this master password automatically to decrypt your credentials.
```

OS X Keystore Integration::

> Available on macOS. This provider uses a random value that is stored in the OS Keychain. You can check it using the Keychain Access utility. This provider uses the master password automatically to decrypt your credentials.

Reviewer: OS Keychain > the iOS Keychain? the iCloud Keychain? an OS keychain?

Reviewer: What do we mean by "check it"? Discover its value? Verify that it exists? Something else?

Reviewer: Third sentence:
```suggestion
OS X Keystore Integration uses the master password automatically to decrypt your credentials.
```

Linux Integration (64 bit)::

> Available in Linux. This provider uses a random value and relies on the `libsecret` library that must be present and working on your Linux distribution. This provider uses the master password automatically to decrypt your credentials.

Reviewer: Does the provider or the random value "rely" on the library?
What do we mean by "rely": does the value come from the library, or is it based on something in that library, or...what? 
Ditto naming provider instead of "This provider" in the third sentence.

Author: It's based on something in the library. The point of using "rely" is that it the library is not present, this provider does not work.

UI Prompt::
Available in Windows, macOS, and Linux. This provider prompts you to create a master password and uses it to encrypt your credentials. This provider requires you to enter your master password every time you open Studio.

> 

Reviewer: See reorg suggestion above for the lines remaining in this topic.
What follows are copyedits of existing text only.

> When adding a new credential for the first time, Studio notifies you that a new master password has been created. Additionally, Studio prompts you to create two password hints to use in case you forget your master password. The password hints are security questions that you must treat as a second password. If you lose your master password, your password hints are the only other method to recover your credentials.

Reviewer: Who or what is adding the new credential?
If Studio, consider "When Studio adds..., it notifies you....

Author: Yes, but I could use this to explain that by providing these hints a user may be able to recover an overwritten former master password, I'm basing this on Eclipse's help around this:

> Lost secure storage password or invalid secure storage password?
If you have specified password recovery questions/answers, you can use the command link General > Security > Secure Storage preference page to recover the master password. (...)  The recovered password is not displayed, but stored in the memory cache until the application is closed.

Based on the "invalid secure storage password" part, I assume this might be a good workaround for a user whose password was updated by another session. @mdaloia Is that correct?

Reviewer 2: That's correct.

The questions are saved in the secure_storage file and the master password for a particular provider is encrypted with the combination of the answers of both questions. As the master password is saved encrypted inside the secure_storage file this is the only mechanism that allows to recover it, as in the case you described or, for example, if you make a backup of this file and want to use it after reinstalling your O.S.

Each provider supplies its own master password. When you use a provider to encrypt a new credential, you can only use that provider to recover the encrypted credential.

> Once a master password is created, Studio caches it until it is closed, or you manually clear the password from Studio's cache.

Reviewer: ```suggestion
After a master password is created, Studio caches it until it is closed, or you manually clear the password from Studio's cache.
```
Please clarify the second "it": Studio caches the password until <the password or Studio?> is closed.

== Secure Storage File

This file holds all your user credentials in an encrypted form. By default, the secure storage file is located in `$HOME/.eclipse/org.eclipse.equinox.security/secure_storage`.

> Because it is a single file, the same secure storage is shared across all Eclipse-based workspaces and Anypoint Studio instances.

Reviewer: ```suggestion
This file holds all your user credentials in an encrypted form. By default, the secure storage file is located in `$HOME/.eclipse/org.eclipse.equinox.security/secure_storage` and is shared across all Eclipse-based workspaces and Anypoint Studio instances.
```

== Update the Master Password

> You can change or rotate the master password for a particular provider. When changing a master password, all the encrypted data by this provider is re-encrypted with the new master password.

Reviewer: rotate? Do you mean set up a password rotation rather than make a one-time change?
Also...under what circumstances would your audience want or need to do this?
And I suggest clarifying who or what is changing the master password, just because I think I initially misunderstood this earlier: "When you change a master password, all...."
```suggestion
If <this happens, or some other reason to change>, you can either make a one-time master password change or establish an ongoing master password change rotation. When you change a master password, all data encrypted by your provider is automatically reencrypted with the new master password.
```

Author: Good point: 'change' and 'rotate' here are referring to the same thing. I'll stick with 'change' to avoid. confusion.

=== Before you Begin

> * Ensure that all other instances of Anypoint Studio and any other Eclipse-based application are not running. +
> Because the secure storage file is shared across all running instances and the master password is cached in each instance's memory, updating your master password causes other opened instances to still write the credentials using the old master password. Closing all other instances ensures you avoid decryption issues. See <<errors, Decryption Errors>> for more information.

Reviewer: I think of prerequisites as WHAT is required, not WHY, which is conceptual information.
That said, I would make this a one-sentence text:
```suggestion
To avoid decryption issues, ensure that all other instances of Anypoint Studio and any other Eclipse-based application are not running. +
```

> To re-generate your master password:

Reviewer: style ref, generally no hyphen following prefixes:
```suggestion
To regenerate your master password:
```

. Navigate to *Preferences* > *General* > *Security* > *Secure Storage*.
. Select *Change Password...*:
+
image::secure-storage-change-password.png[]

> . Studio prompts you to create a password hint for this new password. +
> Select *Yes* to generate a new password hint:

Reviewer: ```suggestion
. When Studio prompts you to create password hints, select *Yes*:
```
two hints, yes?

+
image::new-password-created.png[]
+
[CAUTION]
--

> If you don't supply new password hints, then the former ones persist, which reference the former password. You cannot recover this new password using the previous password hints.

Reviewer 2: The usage of "password hint" it is really misleading from Eclipse side, they are really "Recovery questions/answers". I understand that this is to be close to UI messages but maybe it could be clarified.

Also, on each reference to "password" of this block I would make explicitly that it is the master password that it is going to be changed.

Author: Good point. I want to refer to them as "password hints" because of the UI. 
We always documented Studio as if the user is not familiar with Eclipse, so it's preferable to stick with the definitions that are presented to the user in Studio. 
In the first mention of "password hint" I'll add a clarification that these are "security questions".

> If you don't supply new password hints, then the former ones persist, which reference the former master password. You cannot recover this new master password using the previous password hints.

Reviewer: ```suggestion
If you don't supply new password hints, then the existing hints persist, referencing the existing master password, and you will not be able to recover your new master password.
```
Hmm...this sounds like a BUG to me: the process should not allow you to proceed without creating the two new hints and verifying that they ARE new....

--
. Studio prompts you that the password has been successfully updated.

> . Select *Apply and Close*.

Reviewer: As with Line 54 above, Line 63 is not a step, so we have to reword:
```suggestion
. When Studio confirms that the password is successfully updated, select *Apply and Close*.
```

Reviewer: BREAK here to commit.

== Customize the Location of the Secure Storage and Password Files

> To change the location of the password and secure storage files:

Reviewer: Avoid simply repeating a topic title in its first sentence. Maybe add something about why your audience might want to do this, under what conditions.

. Locate your `AnypointStudio.ini` file:
* In Windows:

> .. Go to your Anypoint Studio's installation directory.

Reviewer: ```suggestion
.. Go to your Anypoint Studio installation directory.
```

* In macOS:

> .. Go to your `/Applications` folder, right-click the `AnypointStudio` package and select *Show Package Contents*.

Reviewer: ```suggestion
.. Go to your `/Applications` folder, right-click the `AnypointStudio` package, and select *Show Package Contents*.
```

.. Go to `Contents` > `Eclipse`, and open the `AnypointStudio.ini` file.
* In Linux:
.. Go to `/usr/lib/anypointstudio/anypointstudio.ini`.

> . Add the following options together to bypass the password provider mechanism:

Reviewer: together > at one time? without taking any action in between? or...do we need "together" at all?

+
[source]
--
-eclipse.keyring <file_path> //<1>
-eclipse.password <file_path> //<2>
--
<1> Configures the location of the Secure Storage file.
<2> Configures the location of the password file. Studio uses this file's content as the master password.

> 

Reviewer 2: Was the omission of the another alternative (via arguments in the command line) on purpose?

> <2> Configures the location of the password file. Studio uses this file's content as the master password.

Reviewer: ```suggestion
<2> Configures the location of the password file, which Studio uses as the master password.
```

Reviewer: Is there no action after adding the options?

[CAUTION]
You can only decrypt your credentials with the same password file used to encrypt them.

[[errors]]
== Decryption Errors

> 

Reviewer: Did you mean these to be reversed?
Doesn't LIne 90 result in an error that you want to doc after Line 93?
Also, you typically do not want to have two contiguous titles, so...can you think of a transitional text between the Decryption Errors title and the subtitle of the first error described?

[[no-password-provided-exception]]
=== No Password Provided

Exception: `org.eclipse.equinox.security.storage.StorageException: No password provided`

> Studio throws this error when it loads and caches the master password for the provider and reads a password in the secure storage that cannot decrypt.

Reviewer 2: Suggestion as I found it confusing on how it was worded:

Studio throws this error when it loads and caches the master password for the provider and then it tries to decrypt a value that was encrypted with another master password.

> This error occurs when Studio loads and caches the master password for the provider and then it tries to decrypt a value that was encrypted with another master password.

Reviewer: What is "it": Studio or the provider? Please clarify by replacing "it".

> This happens when one instance of Anypoint Studio changes the master password while another instance of Studio saves a new credential using a cached master password.

Reviewer 2: Maybe clarify:
... one instance of Anypoint Studio (or any Eclipse based application) ...

Also, I would clarify that Studio does not change the master password by itself. It is a user action.

> This happens when a user changes the master password using one instance of Studio or another Eclipse-based application changes the master password while another instance of Studio saves a new credential using a cached master password.

Reviewer: Looking at the following error, I see that the two texts begin in nearly the same way. 
So should this one actually be:
```suggestion
This happens when a user of one instance of Studio or another Eclipse-based application changes the master password at the same time that a user in another instance of Studio saves a new credential using the previously cached master password.
```

Reviewer: Do we need both the texts at Line 99 and Line 101: are they documenting the same reason, just using different words? I don't understand how the two texts relate to one another, but maybe your audience will? Please evaluate for clarity; maybe get a writing peer's advice?

> This happens when a user changes the master password using one instance of Studio or another Eclipse-based application changes the master password while another instance of Studio saves a new credential using a cached master password.

Reviewer: Do we need both the texts at Line 99 and Line 101: are they documenting the same reason, just using different words?

=== Given Final Block Not Properly Padded

`org.eclipse.equinox.security.storage.StorageException: Given final block not properly padded`.

> Studio throws this error when it attempts to use a bad key during decryption.

Reviewer 2: Was on purpose to not add more information about this error like the previous one?

This line is base on the actual full error message that is thrown (which on the title was shortened here). It could confuse the user about the mention of "key" which is an internal name and even it is not the same I think it is better to write it as "master password".

Maybe I would explain better why it is produced and when this could happen.

Author: You are right. I didn't notice that I didn't complete this description. 

I just pushed an update. 79bf1da

> This error occurs when the current cached master password can not decrypt a value because it was encrypted with another password.

Reviewer: ```suggestion
This error occurs when the currently cached master password cannot decrypt a value because that value was encrypted using the previously cached master password.
```

Reviewer: As you writer going forward, consider "it" a red-flag word and think about whether you should name what "it" actually is.

> This happens when a user using an instance of Studio or another Eclipse-based application changes the master password and another instances attempts to read encrypted values with a previous cached password.

Reviewer: ```suggestion
This happens when a user using one instance of Studio or another Eclipse-based application changes the master password, and another instance of Studio or the application attempts to read encrypted values by using the previous cached master password.
```

Reviewer: ditto: do we need both 108 and 110?
